src/main/java/core/di/factory/BeanFactory.java

```diff
@@ -37,46 +49,40 @@ public void initialize() {
     }
 
     public Set<Class<?>> getControllers() {
-        return annotatedTypes.stream()
+        return scannedAnnotatedTypes.stream()
                 .filter(type -> type.isAnnotationPresent(Controller.class))
                 .collect(Collectors.toSet());
     }
 
-    private void createBeans(Class<? extends Annotation> annotation) {
-        annotatedTypes.stream()
-                .filter(bean -> bean.isAnnotationPresent(annotation))
-                .forEach(ExceptionWrapper.consumer(
-                        bean -> {
-                            final Constructor<?> injectedConstructor = getInjectedConstructor(bean);
-                            final Object[] parameterBeans = getParameterBeans(injectedConstructor);
-
-                            beans.put(bean, BeanUtils.instantiateClass(injectedConstructor, parameterBeans));
-                        }));
-    }
+    private Object instanticate(Class<?> type) {
+        // Bean 저장소에 clazz에 해당하는 인스턴스가 이미 존재하면 해당 인스턴스 반환
+        final Object found = beans.get(type);
+        if (found != null) {
+            return found;
+        }
 
-    private Constructor<?> getInjectedConstructor(Class<?> bean) throws NoSuchMethodException {
-        final Constructor<?> injectedConstructor = Arrays.stream(bean.getDeclaredConstructors())
-                .filter(constructor -> constructor.isAnnotationPresent(Inject.class))
-                .max(Comparator.comparing(Constructor::getParameterCount))
-                .orElse(null);
+        // clazz에 @Inject가 설정되어 있는 생성자를 찾는다. BeanFactoryUtils 활용
+        final Constructor<?> injectedConstructor = BeanFactoryUtils.getInjectedConstructor(type);
 
+        // @Inject로 설정한 생성자가 없으면 Default 생성자로 인스턴스 생성 후 Bean 저장소에 추가 후 반환
         if (injectedConstructor == null) {
-            return bean.getDeclaredConstructor();
+            return instantiateClass(type);
         }
 
-        return injectedConstructor;
+        // @Inject로 설정한 생성자가 있으면 찾은 생성자를 활용해 인스턴스 생성 후 Bean 저장소에 추가 후 반환
+        return instantiateConstructor(injectedConstructor);
     }
 
-    private Object[] getParameterBeans(Constructor<?> injectedConstructor) {
-        return Arrays.stream(injectedConstructor.getParameterTypes())
-                .map(ExceptionWrapper.function(this::getObjectFromBeans))
-                .toArray();
+    private Object instantiateClass(Class<?> type) {
+        return BeanUtils.instantiateClass(type);
     }
 
-    private Object getObjectFromBeans(Class<?> parameter) throws ClassNotFoundException {
-        return beans.values().stream()
-                .filter(value -> parameter.isAssignableFrom(value.getClass()))
-                .findFirst()
-                .orElseThrow(ClassNotFoundException::new);
+    private Object instantiateConstructor(Constructor<?> constructor) {
+        final Object[] parameters = Arrays.stream(constructor.getParameterTypes())
+                .map(ExceptionWrapper.function(
```

Comment:
인자가 인터페이스 일 때 그 구현 클래스를 넘겨줄 수 있도록 아래 코드가 추가되어야할 것 같아요.
```suggestion
                .map(parameterType -> BeanFactoryUtils.findConcreteClass(parameterType, scannedAnnotatedTypes))
                .map(ExceptionWrapper.function(
```